nixos/modules/security/vault/keys.nix

```diff
@@ -0,0 +1,420 @@
+    ))));
+
+    systemd.services = # create a service for every key
+      (listToAttrs (flip mapAttrsToList finalKeys (_: key: {
+        name = "vault-key-${key.name}";
+        value =
+          let
+            # the default template
+            default = head key.templates;
+
+            keyRenderWait = pkgs.writeScriptBin "key-render-wait" (builtins.readFile ./key-render-wait.py);
+
+            # filenames for the temporary files and target sink files
+            tempFileNames = concatStringsSep "," (map (template: ".${template.id}.tmp") key.templates);
+            sinkFileNames = concatStringsSep "," (flatten (flip mapAttrsToList key.sinks (_: sink:
+              if (key.template == null && ((length key.templates) > 1)) then (map (t: "${sink.name}-${t.suffix}") key.templates) else sink.name
+            )));
+          in
+          {
+            description = "Vault key activation service";
+
+            before = key.dependency.units;
+            after = [ "vault-agent.target" ];
+            wants = [ "vault-agent.target" ];
+            wantedBy = [ "multi-user.target" ] ++ key.dependency.units;
+
+            path = with pkgs; [
+              (python3.withPackages (p: with p; [
+                inotify-simple
+              ]))
+            ];
+
+            serviceConfig = {
+              Restart = "always";
+              RestartSec = "100ms";
+              TimeoutStartSec = "infinity";
+            };
+
+            preStart = ''
+              # Wait for create/move if any keys don't exist here
+              ${keyRenderWait}/bin/key-render-wait -t "${tempFileNames}" -s "${sinkFileNames}" -d "${key.folder}"
+            '';
+
+            script = ''
+              set -euo pipefail
+
+              # if the key's folder does not exist, create it
+              if [[ ! -d "${key.folder}" ]]; then
+                  mkdir -p "${key.folder}"
+                  chmod -R 0600 "${key.folder}"
+              fi
+
+              # we are now DONE waiting for all the keys
+              # finally render them
+            ${concatStringsSep "\n" (flatten ([
+              (flip mapAttrsToList key.sinks (_: sink:
+              let
+              buildKey = render: dest: id:
+              let
+                  tempFile = "${key.folder}/.${id}.tmp";
+              in
+              # if the temp file doesn't exist, don't attempt to render and instead use the old version
+              ''
+                if [[ -f "${tempFile}" ]]; then
+                    rm -rf "${dest}"
+                    ${render}
+                    chmod ${sink.permissions} "${dest}"
+                    chown "${sink.user}:${sink.group}" "${dest}"
+                fi
+              '';
+
+              buildStandaloneKey =
+              let
+                  dest = "${key.folder}/${sink.name}";
+              in
```

Comment:
Why do you need this `in`?

Reply:
(I am trying to make sense of the whole order, it is a bit hard, I think, moving stuff in a library or encapsulating logic would help a lot here.)